File: Binary Tree/654.最大二叉树.cpp
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=654 lang=cpp
 *
 * [654] 最大二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    TreeNode *constructMaximumBinaryTree(vector<int> &nums)
    {
        if (nums.size() == 0)
        {
            return nullptr;
        }
        if (nums.size() == 1)
        {
            return new TreeNode(nums[0]);
        }
        int index = 0;
        int middle = nums[0];
        // 找到最大的元素和下标
        for (int i = 1; i < nums.size(); i++)
        {
            if (middle < nums[i])
            {
                middle = nums[i];
                index = i;
            }
        }

        vector<int> left(nums.begin(), nums.begin() + index);
        // 这里+1是为了过滤掉已经是根节点的middle节点
        vector<int> right(nums.begin() + index + 1, nums.end());
        TreeNode *root = new TreeNode(middle);

        root->left = constructMaximumBinaryTree(left);
        root->right = constructMaximumBinaryTree(right);

        return root;
    }
};
// @lc code=end

File: Binary Tree/二叉树的修改与构造/654.最大二叉树.ts
/*
 * @lc app=leetcode.cn id=654 lang=typescript
 *
 * [654] 最大二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function constructMaximumBinaryTree(nums: number[]): TreeNode | null {
  if (nums.length === 0) {
    return null
  }

  if (nums.length === 1) {
    return new TreeNode(nums[0])
  }

  let max = Number.MIN_SAFE_INTEGER
  let index = 0
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] > max) {
      max = nums[i]
      index = i
    }
  }

  let left = nums.slice(0, index)
  let right = nums.slice(index + 1)

  return new TreeNode(max,
    constructMaximumBinaryTree(left),
    constructMaximumBinaryTree(right))

};
// @lc code=end

